File: Data Structures & Algorithms/duplicate-integer/submission-2.cpp
class Solution {
public:
    bool hasDuplicate(vector<int>& nums) {
        unordered_set<int> dups;
        for(int i=0; i<nums.size(); i++){
            if(dups.contains(nums[i])){
                return true;
            }
            dups.insert(nums[i]);
        }
        return false;
    }
};

File: Data Structures & Algorithms/subsets-ii/submission-2.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()                       # canonical order so dups collapse
        seen = set()                      # real set, not {}
        sub = []
        def dfs(i):
            if i >= len(nums):
                seen.add(tuple(sub))      # tuple: hashable + snapshot
                return
            sub.append(nums[i])
            dfs(i + 1)
            sub.pop()
            dfs(i + 1)
        dfs(0)
        return [list(t) for t in seen]    # tuples back to lists for output